x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/get_subfield_changes.ts

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type {
+  RuleKqlQuery,
+  RuleDataSource,
+  RuleEqlQuery,
+  RuleEsqlQuery,
+  InlineKqlQuery,
+  RuleSchedule,
+  RuleNameOverrideObject,
+  TimestampOverrideObject,
+  TimelineTemplateReference,
+  BuildingBlockObject,
+  Threshold,
+  DiffableAllFields,
+} from '../../../../../../../common/api/detection_engine';
+import {
+  sortAndStringifyJson,
+  getSubfieldChangesForDataSource,
+  getSubfieldChangesForKqlQuery,
+  getSubfieldChangesForEqlQuery,
+  getSubfieldChangesForRuleSchedule,
+  getSubfieldChangesForRuleNameOverride,
+  getSubfieldChangesForTimestampOverride,
+  getSubfieldChangesForTimelineTemplate,
+  getSubfieldChangesForBuildingBlock,
+  getSubfieldChangesForThreshold,
+  getSubfieldChangesForEsqlQuery,
+  getSubfieldChangesForThreatQuery,
+} from './get_subfield_changes_for_field';
+import type { SubfieldChange } from './types';
+
+/**
+ * Splits a field into subfields and returns the changes between the old and new subfield values.
+ *
+ * @param fieldName - The name of the field for which subfield changes are to be computed.
+ * @param oldFieldValue - The old value of the field.
+ * @param newFieldValue - The new value of the field.
+ * @returns - An array of subfield changes.
+ */
+export const getSubfieldChanges = <FieldName extends keyof DiffableAllFields>(
+  fieldName: FieldName,
+  oldFieldValue?: DiffableAllFields[FieldName],
+  newFieldValue?: DiffableAllFields[FieldName]
+): SubfieldChange[] => {
+  switch (fieldName) {
+    case 'data_source':
+      return getSubfieldChangesForDataSource(
+        oldFieldValue as RuleDataSource | undefined,
+        newFieldValue as RuleDataSource | undefined
```

Comment:
Def agree this isn't the best approach. I couldn't get it to work quickly without typecasting. I would appreciate it if we could pair and improve the types here.

Reply:
Discussed this offline. It turned out it's not possible to avoid typecasting since TS doesn't support narrowing union types. Added an explainer comment in code. We may want to discuss the type situation at the upcoming Tech Time.